### DISCUSSION TITLE
Use `svcs` to fill in arguments of a function call or a type (unregistered) instantiation.

### DISCUSSION THREAD

**Discussion by jbasko** [hynek/svcs#79] [Q&A]
Can I use `svcs` to inject dependencies in callables based on annotations as I am trying to in the following example:

```python

import pytest
import svcs


@pytest.fixture()
def registry():
    return svcs.Registry()


@pytest.fixture()
def container(registry):
    return svcs.Container(registry=registry)


def test_can_call_function_with_deps_injected(registry, container):
    def doubles_int(x: int):
        return x * 2

    registry.register_value(int, 5)
    assert container.get(doubles_int)() == 10

```

Similarly, when I create an instance of an unregistered type which depends on some registered stuff, from a DI framework I would expect a utility which injects it so that I don't have to explicitly request them from the container. Is there a thing like that in `svcs`?

**Comment by hynek** [comment]
Are you asking for some kind of `container.call_and_inject(f)` that introspects the signature of `f` and then, according to their types presumably, runs the appropriate `container.get()` and passes the results as arguments?

**Reply by jbasko** [reply]
Yes, exactly. But now I've read a bit more, I see that you recommend `incant` for that.

**Reply by hynek** [reply]
True, it might be a fun exercise to implement it on top of *svcs* tho. I think @Tinche even did it once.

**Reply by hynek** [reply]
I _think_ #94 might be an interesting compromise?

**Reply by jbasko** [reply]
Kind of, it could be a usable building block. Thanks.

I would not want to register every function, and also I would expect that this does not cache the results (so, not singletons), and that it does not expect all arguments to be injectables (instead, some can be passed at the time when one calls `inspect_and_get`). I was after `inspect_and_call`, but I would expect something like `inspect_and_get` to be available as well.

**Reply by hynek** [reply]
only two years later, it's live in svcs 26.1 as auto wiring: https://svcs.hynek.me/en/stable/autowiring.html

do you think your issue could be achieved using that?
